Fix contract interactions error

rin-st opened the pull request:
See the bug and how to reproduce in #1145 

It can't parse the error because it checks only abi of the ContractA. This PR fixes it, now when trying to find error it parses abi's of all the deployed contracts.

Error after fix
<img width="444" height="127" alt="image" src="https://github.com/user-attachments/assets/ad4b5373-7844-47f8-8e06-046b57b55832" />


Fixes #1145

Pabl0cks reviewed (approved):
Tested with token vendor challenge, trying to sell more tokens than current user balance, and it's working nicely to me.

Before vs After:

<img width="576" height="185" alt="image" src="https://github.com/user-attachments/assets/c4736088-60ca-4d19-87ce-fd112e1edc98" />

technophile-04 reviewed (commented):
Thanks @rin-st and Pablo for testing!! Looks nice to me! Just added a small comment 🙌 

Ohh also update if you do `yarn next:check-types` without any contract deployment you get: 

<img width="1506" height="297" alt="image" src="https://github.com/user-attachments/assets/f7604ee9-0918-4b6c-8d80-72a5fb39549c" />

technophile-04 commented on packages/nextjs/utils/scaffold-eth/contract.ts:
```diff
@@ -335,6 +337,68 @@ export type UseScaffoldEventHistoryData<
+    const signatureMatch = originalParsedError.match(/0x[a-fA-F0-9]{8}/);
+    const signature = signatureMatch ? signatureMatch[0] : "";
+
+    if (signature) {
```
Maybe we could early returns instead of nested if blocks 

if (!signature) return originalParsedError;

rin-st replied:
[9f2e17e](https://github.com/scaffold-eth/scaffold-eth-2/pull/1146/commits/9f2e17e35e9e7de24d9ce4687c800642b3b6078d)

technophile-04 commented on packages/nextjs/utils/scaffold-eth/contract.ts:
```diff
@@ -335,6 +337,68 @@ export type UseScaffoldEventHistoryData<
+      try {
+        // Get all deployed contracts for the current chain
+        const chainContracts = deployedContractsData[chainId];
+        if (chainContracts) {
```
Same here early return: 

`if (!chainContracts) return;`

rin-st replied:
[9f2e17e](https://github.com/scaffold-eth/scaffold-eth-2/pull/1146/commits/9f2e17e35e9e7de24d9ce4687c800642b3b6078d)

rin-st commented on packages/nextjs/utils/scaffold-eth/contract.ts:
```diff
@@ -345,7 +409,9 @@ export const simulateContractWriteAndNotifyError = async ({
     await simulateContract(wagmiConfig, params);
   } catch (error) {
-    const parsedError = getParsedError(error);
+    const chainId = wagmiConfig.chains[0]?.id as AllowedChainIds;
```
[4a16ecd](https://github.com/scaffold-eth/scaffold-eth-2/pull/1146/commits/4a16ecdc13c3199aec90dbd5ef26db2e3fa711a5) Updated, thank you!

rin-st commented:
Thanks for reviews, guys!

> Ohh also update if you do yarn next:check-types without any contract deployment you get:

Fixed! [33fa96f](https://github.com/scaffold-eth/scaffold-eth-2/pull/1146/commits/33fa96fb4564e0459a0cb0aaf8af6481c83856fe)

---

One more note: if we disable simulate
```
const { writeContractAsync: writeContractAAsync } = useScaffoldWriteContract({
    contractName: "ContractA",
    disableSimulate: true,
  })
```
we're getting the same error as before.

So I updated `useTransactor` a bit so it also parses all the abi's [38fd747](https://github.com/scaffold-eth/scaffold-eth-2/pull/1146/commits/38fd7477247c286bf80320ddedb4c69366cf8be0)

technophile-04 commented:
Tysm Rinat!! Looks great! Just pushed a small commit, instead of using hardcode hardaht chainId used the scaffoldConfig's 0th network in `useTransactor` since contract on 0th chainId is assumed to be on all chains. I mean this case is not going to happen more likely because walletClient with give `chainId` but just incase 🙌

damianmarti commented:
This is great @rin-st!! Nice improvement!
